src/relay/dnsrelay/mod.rs

```diff
@@ -0,0 +1,260 @@
+use std::{
+    future::Future,
+    net::{IpAddr, SocketAddr},
+    time::Duration,
+};
+
+use tokio::{
+    io::Result,
+    net::{TcpStream, UdpSocket},
+    prelude::*,
+    time,
+};
+
+use log::debug;
+
+use trust_dns_proto::{op::*, rr::*};
+
+use crate::{context::SharedContext, relay::socks5::*};
+
+pub async fn try_timeout<T, F>(fut: F, timeout: Option<Duration>) -> io::Result<T>
+where
+    F: Future<Output = Result<T>>,
+{
+    match timeout {
+        Some(t) => time::timeout(t, fut).await?,
+        None => fut.await,
+    }
+    .map_err(From::from)
+}
+
+async fn udp_lookup(qname: &Name, qtype: RecordType, server: SocketAddr) -> Result<Message> {
+    let mut socket = UdpSocket::bind(("0.0.0.0", 0)).await?;
+
+    let mut message = Message::new();
+    let mut query = Query::new();
+
+    query.set_query_type(qtype);
+    query.set_name(qname.clone());
+
+    message.set_id(6666);
+    message.set_recursion_desired(true);
+    message.add_query(query);
+
+    let req_buffer = message.to_vec()?;
+    socket.send_to(&req_buffer, server).await?;
+
+    let mut res_buffer = vec![0; 512];
+    socket.recv_from(&mut res_buffer).await?;
+
+    Ok(Message::from_vec(&mut res_buffer)?)
+}
+
+async fn socks5_lookup(qname: &Name, qtype: RecordType, socks5: SocketAddr, ns: SocketAddr) -> Result<Message> {
+    let mut stream = TcpStream::connect(socks5).await?;
+
+    // 1. Handshake
+    let hs = HandshakeRequest::new(vec![SOCKS5_AUTH_METHOD_NONE]);
+    hs.write_to(&mut stream).await?;
+    stream.flush().await?;
+
+    let hsp = HandshakeResponse::read_from(&mut stream).await?;
+    assert_eq!(hsp.chosen_method, SOCKS5_AUTH_METHOD_NONE);
```

Comment:
I think it is okay to send the entire packet at once, which is what `LocalDnsServer` does in shadowsocks-android.

Reply:
Also lower priority...